### PR TITLE
sql: decode failed `CPut`s' keys instead of tracking writes

### PR DESCRIPTION
**Pull request by tamird**

### CODE REVIEW

File: sql/table.go [hunk 1]
```diff
@@ -251,21 +269,19 @@ func makeKeyVals(desc *TableDescriptor, columnIDs []ColumnID) ([]parser.Datum, e
...
-	var tableID uint64
-	key, tableID = encoding.DecodeUvarint(key)
-	if ID(tableID) != desc.ID {
-		return nil, fmt.Errorf("%s: unexpected table ID: %d != %d", desc.Name, desc.ID, tableID)
+	var (
+		indexID IndexID
+		err     error
+	)
+
+	indexID, key, err = decodeIndexKeyPrefix(desc, key)
```
Comment (petermattis) [inline sql/table.go]: Can't you use `:=` here instead of declaring the variables above?

Comment (tamird) [inline sql/table.go]: eh, yeah, slight rearrangement

File: sql/errors.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+		valStrs = append(valStrs, val.String())
+	}
+
+	return fmt.Sprintf("duplicate key value (%s)=(%s) violates unique constraint %q",
+		strings.Join(e.index.ColumnNames, ","),
+		strings.Join(valStrs, ","),
+		e.index.Name)
+}
+
+func runBatchWithErrorConversion(tableDesc *TableDescriptor, runBatch func(*client.Batch) error, b client.Batch) error {
```
Comment (petermattis) [inline sql/errors.go]: Minor style preference to leave the running of the batch in the caller and to rename this function to something like `convertBatchError(b client.Batch, err error)`.

File: sql/errors.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+					valTypes, err := makeKeyVals(tableDesc, index.ColumnIDs)
+					if err != nil {
+						return err
+					}
+					vals := make([]parser.Datum, len(valTypes))
+					if _, err := decodeKeyVals(valTypes, vals, key); err != nil {
+						return err
+					}
+
+					return errUniquenessConstraintViolation{index: index, vals: vals}
```
Comment (petermattis) [inline sql/errors.go]: Was there a value to creating the custom error type here vs calling `fmt.Errorf`? Just wondering if I'm missing something about how it is being used.

**Comment by petermattis** [comment]
LGTM modulo comments. So much simpler.

File: sql/errors.go (hunk 3 above)
Comment (tamird) [inline sql/errors.go]: nah, just attempting to add a bit of structure

Comment (tamird) [inline sql/errors.go]: (and to justify the creation of errors.go)

File: sql/errors.go (hunk 2 above)
Comment (tamird) [inline sql/errors.go]: gonna exercise artistic license on this one

**Comment by spencerkimball** [comment]
Btw, I think you should be more explicit in your PR descriptions. At this point, I'm struggling trying to keep up with everything without reading all of the code changes and I can't determine what exactly this PR does short of reading diffs. Try to write your PR title and description for an audience a little wider than just your reviewer.

**Comment by tamird** [comment]
Sorry about that. This PR changes the mechanism by which we generate error message on uniqueness violations. Previously, we were bookeeping using function-local state, while now we are generating the error messages by decoding the failed CPut key. No functional change.

**Pull request merged**